server/client.go

```diff
@@ -3324,39 +3334,95 @@ var needFlush = struct{}{}
+	// Check if message tracing is enabled.
+	mt, mtd := c.isMsgTraceEnabled()
+
+	client := sub.client
 	// Check sub client and check echo. Only do this if not a service import.
-	if sub.client == nil || (c == sub.client && !sub.client.echo && !sub.si) {
+	if client == nil || (c == client && !client.echo && !sub.si) {
+		if client != nil && mt != nil {
+			client.mu.Lock()
+			mt.addEgressEvent(client, sub, errMsgTraceNoEcho)
+			client.mu.Unlock()
+		}
 		return false
 	}
 
-	client := sub.client
 	client.mu.Lock()
 
 	// Check if we have a subscribe deny clause. This will trigger us to check the subject
 	// for a match against the denied subjects.
 	if client.mperms != nil && client.checkDenySub(string(subject)) {
+		if mt != nil {
+			mt.addEgressEvent(client, sub, errMsgTraceSubDeny)
+		}
 		client.mu.Unlock()
 		return false
 	}
 
 	// New race detector forces this now.
 	if sub.isClosed() {
+		if mt != nil {
+			mt.addEgressEvent(client, sub, errMsgTraceSubClosed)
+		}
 		client.mu.Unlock()
 		return false
 	}
 
 	// Check if we are a leafnode and have perms to check.
 	if client.kind == LEAF && client.perms != nil {
 		if !client.pubAllowedFullCheck(string(subject), true, true) {
+			if mt != nil {
+				mt.addEgressEvent(client, sub, errMsgTracePubViolation)
+			}
 			client.mu.Unlock()
 			client.Debugf("Not permitted to deliver to %q", subject)
 			return false
 		}
 	}
 
+	var mtErr string
+	if mt != nil {
+		// For non internal subscription, and if the remote does not support
+		// the tracing feature...
+		if sub.icb == nil && !client.msgTraceSupport() {
+			// If we are doing delivery, we will still forward the message,
+			// but we add an error to the Egress event to hint that one should
+			// not expect a tracing event from that remote.
+			if mtd {
+				mtErr = errMsgTraceNoSupport
+			} else {
+				// We are not sending the message at all because the user
+				// expects a trace-only and the remote does not support
+				// tracing, which means that it would process/deliver this
+				// message, which may break applications.
+				// Add the Egress with the no-support error message.
+				mt.addEgressEvent(client, sub, errMsgTraceOnlyNoSupport)
+				client.mu.Unlock()
+				return false
+			}
+		}
+		// For ROUTER, GATEWAY and LEAF, even if we intend to do tracing only,
+		// we will still deliver the message. The remote side will
+		// generate an event based on what happened on that server.
+		if !mtd && (client.kind == ROUTER || client.kind == GATEWAY || client.kind == LEAF) {
+			mtd = true
+		}
+		// If we skip delivery and this is not for a service import, we are done.
+		if !mtd && (sub.icb == nil || c.noIcb) {
+			mt.addEgressEvent(client, sub, _EMPTY_)
+			client.mu.Unlock()
+			// Although the message is not actually delivered, for the
+			// purpose of "didDeliver", we need to return "true" here.
+			return true
+		}
+	}
+
 	srv := client.srv
 
-	sub.nm++
+	if mt == nil || mtd {
```

Comment:
Why would we qualify this here?

Reply:
Because we don't want to update that if we were doing a trace only. I have changed isMsgTraceEnabled to actually return if we are doing trace only (which then is false when mt is nil or when mt is not but we are doing message delivery). I also added a comment to make it clear as to why now.